backend/src/schema/resolvers/emails.js

```diff
@@ -0,0 +1,92 @@
+import generateNonce from './helpers/generateNonce'
+import Resolver from './helpers/Resolver'
+import existingEmailAddress from './helpers/existingEmailAddress'
+import { UserInputError } from 'apollo-server'
+import Validator from 'neode/build/Services/Validator.js'
+
+export default {
+  Mutation: {
+    AddEmailAddress: async (_parent, args, context, _resolveInfo) => {
+      let response
+      try {
+        const { neode } = context
+        await new Validator(neode, neode.model('UnverifiedEmailAddress'), args)
+      } catch (e) {
+        throw new UserInputError('must be a valid email')
+      }
+
+      // check email does not belong to anybody
+      await existingEmailAddress(_parent, args, context)
+
+      const nonce = generateNonce()
+      const {
+        user: { id: userId },
+      } = context
+      const { email } = args
+      const session = context.driver.session()
+      const writeTxResultPromise = session.writeTransaction(async txc => {
+        const result = await txc.run(
+          `
+            MATCH (user:User {id: $userId})
+            MERGE (user)<-[:BELONGS_TO]-(email:UnverifiedEmailAddress {email: $email, nonce: $nonce})
+            SET email.createdAt = toString(datetime())
+            RETURN email, user
+          `,
+          { userId, email, nonce },
+        )
+        return result.records.map(record => ({
+          name: record.get('user').properties.name,
+          ...record.get('email').properties,
+        }))
+      })
+      try {
+        const txResult = await writeTxResultPromise
+        response = txResult[0]
+      } finally {
+        session.close()
+      }
+      return response
+    },
+    VerifyEmailAddress: async (_parent, args, context, _resolveInfo) => {
+      let response
+      const {
+        user: { id: userId },
+      } = context
+      const { nonce, email } = args
+      const session = context.driver.session()
+      const writeTxResultPromise = session.writeTransaction(async txc => {
+        const result = await txc.run(
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T09:12:35Z" title="Wednesday, October 2nd 2019, 11:12:35 am +02:00">Oct 2, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,92 @@
+import generateNonce from './helpers/generateNonce'
+import Resolver from './helpers/Resolver'
+import existingEmailAddress from './helpers/existingEmailAddress'
+import { UserInputError } from 'apollo-server'
+import Validator from 'neode/build/Services/Validator.js'
+
+export default {
+  Mutation: {
+    AddEmailAddress: async (_parent, args, context, _resolveInfo) => {
+      let response
+      try {
+        const { neode } = context
+        await new Validator(neode, neode.model('UnverifiedEmailAddress'), args)
+      } catch (e) {
+        throw new UserInputError('must be a valid email')
+      }
+
+      // check email does not belong to anybody
+      await existingEmailAddress(_parent, args, context)
+
+      const nonce = generateNonce()
+      const {
+        user: { id: userId },
+      } = context
+      const { email } = args
+      const session = context.driver.session()
+      const writeTxResultPromise = session.writeTransaction(async txc => {
+        const result = await txc.run(
+          `
+            MATCH (user:User {id: $userId})
+            MERGE (user)<-[:BELONGS_TO]-(email:UnverifiedEmailAddress {email: $email, nonce: $nonce})
+            SET email.createdAt = toString(datetime())
+            RETURN email, user
+          `,
+          { userId, email, nonce },
+        )
+        return result.records.map(record => ({
+          name: record.get('user').properties.name,
+          ...record.get('email').properties,
+        }))
+      })
+      try {
+        const txResult = await writeTxResultPromise
+        response = txResult[0]
+      } finally {
+        session.close()
+      }
+      return response
+    },
+    VerifyEmailAddress: async (_parent, args, context, _resolveInfo) => {
+      let response
+      const {
+        user: { id: userId },
+      } = context
+      const { nonce, email } = args
+      const session = context.driver.session()
+      const writeTxResultPromise = session.writeTransaction(async txc => {
+        const result = await txc.run(
```


not a fan of this naming convention... it's as if we don't know what the result of this transaction will be
we have been using `transactionRes`, which is maybe slightly better, but actually still makes it seem like it's a mystery what it returns... 
In this case it's returning the `emailNode`, no?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-02T13:27:07Z" title="Wednesday, October 2nd 2019, 3:27:07 pm +02:00">Oct 2, 2019</time>_

---


`const result` would be a result set. It's a local variable to store that stuff we need for the return value :laughing:

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T14:31:02Z" title="Wednesday, October 2nd 2019, 4:31:02 pm +02:00">Oct 2, 2019</time>_

---


I get that, just think we don't really gain anything by being vague, and it could make our code easier to understand if it's more expressive rather than less.